hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/AbstractLoadPlanBuildingAssociationVisitationStrategy.java

```diff
@@ -640,6 +641,34 @@ public boolean isDuplicateAssociationKey(AssociationKey associationKey) {
 		return fetchedAssociationKeySourceMap.containsKey( associationKey );
 	}
 
+	@Override
+	public boolean isDuplicateAssociatedEntity(AssociationAttributeDefinition attributeDefinition) {
+		if ( attributeDefinition.getAssociationNature() != AssociationAttributeDefinition.AssociationNature.ENTITY ) {
+			return false;  // EARLY EXIT
+		}
+
+		final Joinable currentEntityPersister = (Joinable) currentSource().resolveEntityReference()
+				.getEntityPersister();
+		final AssociationKey currentEntityReferenceAssociationKey =
+				new AssociationKey( currentEntityPersister.getTableName(), currentEntityPersister.getKeyColumnNames() );
+		final AssociationKey associationKey = attributeDefinition.getAssociationKey();
+		final FetchSource registeredFetchSource = registeredFetchSource( associationKey );
+		final EntityReference registeredOwner;
+		if ( registeredFetchSource == null ) {
+			registeredOwner = null;
+		}
+		else if ( registeredFetchSource instanceof Fetch ) {
+			registeredOwner = ( (Fetch) registeredFetchSource ).getSource().resolveEntityReference();
+		}
+		else {
+			registeredOwner = registeredFetchSource.resolveEntityReference();
+		}
+		return isDuplicateAssociationKey( associationKey ) &&
+				( !currentEntityReferenceAssociationKey.equals( attributeDefinition.getAssociationKey() ) ||
+						( registeredOwner != null && !currentEntityPersister.equals( registeredOwner.getEntityPersister() ) ) ||
+						currentEntityPersister.equals( attributeDefinition.toEntityDefinition().getEntityPersister() ) );
```

Comment:
The above logic seems complicated and the format seems not helpful to grasp the overall structure. Maybe joining the first two lines is considerable on this:
```
return ( isDuplicateAssociationKey( associationKey ) && !currentEntityReferenceAssociationKey.equals( attributeDefinition.getAssociationKey() ) ||
						( registeredOwner != null && !currentEntityPersister.equals( registeredOwner.getEntityPersister() ) ) ||
						currentEntityPersister.equals( attributeDefinition.toEntityDefinition().getEntityPersister() ) );
```

Reply:
@NathanQingyangXu, I, personally, don't find your change to be more readable.

I will add some comments to describe the logic, and I will try to simplify it.